poetry/vcs/git.py

```diff
@@ -14,7 +14,8 @@
     "port": r"\d+",
     "path": r"[\w~.\-/\\]+",
     "name": r"[\w~.\-]+",
-    "rev": r"[^@#]+",
+    "rev": r"[^@#?]+",
+    "subdir": r"[\w\-/\\]+",
```

Comment:
A bit unsure about this one. As per the current poetry docuemtnation the argument for `poetry add` when providing a url needs to be a valid git url, which the python vcs dependency url us not. In the curent version we support `<url>#<rev>` (valid git) which is not adherant to pip specification (this seems to be standardised by PEP 610). It would seem that this is not fully compliant to either specification. The use of query string syntax might also vendor dependant (github vs gitlab).

Do you think this could be confusing? And if so, should we pick one and stick with it?

Reply:
When starting with this PR I thought about a new command line argument. I refuse it because it will lead to more changes in the code compared with passing it along with the URL. Second, I was inspired by the way pip is doing this. I guess most new users would find this intuitive.

Other opinions are welcome. I will have to work on this PR the next days, because of the changes introduced by poetry_core now.

Reply:
I like the idea for sure, I am all for consistency with PEP 610. The concern right now is that this change would mean we will neither conform to that nor git's url spec. We should probably make a decision how how to handle this.

More changes might not be a bad thing if it makes things clearer. Additionally, we can definitely tackle exposing this via a cli option in a later PR. Right now, I guess we just need to decide which url spec to follow, we can accept both too - but this would mean some assumptions in error handling during parsing.

There are definitely pros and cons for following either format. From a cli perspective having a standard git url and options that map to the toml spec will reduce the cognitive load on end users as it is explcit (more intuitive?) for someone who is familiar with poetry's config I feel (eg: `poetry add <git url> --subdirectory=x --revision=x` will map cleanly to dependency specification in toml). On the other hand, having it be the same will mean, atleast for new users as you said, be more intuitive.

Reply:
Thinking more about it; we should probably consider enabling both by doing the following.

1. For users who use `poetry add vcs+protocol://repo_url[@rev]/#egg=pkg&subdirectory=pkg_dir` we simply convert this to the appropriate inline table in the toml `pkg = { vcs = "protocol://repo_url", rev = "rev", subdirectory = "pkg_dir"}`. Any cli options `--subdirectory | --revision` overrides values parsed from vcs url. 
2. For users who use `poetry add vcs+protocol://repo_url[#rev]` we maintain current behavior with the possibility of using new cli options.
3. Error out with recommendation if the url is explicitly added to the toml file and is not a valid remote url.

(1) would be really be important for cases where the pip VCS dependency url is copied over.